server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -230,6 +344,142 @@ private void abortMergeTask(MergeTask mergeTask) {
+            }
+            if (leastAvailablePath == null) {
+                LOGGER.error("Cannot read filesystem info");
+                return;
+            }
+            // subtract disk space that already running merges are expected to fill
+            long leastAvailableDiskSpaceBytes = leastAvailablePath.getAvailable().getBytes();
+            for (MergeTask mergeTask : runningMergeTasks) {
+                leastAvailableDiskSpaceBytes -= mergeTask.estimatedRemainingMergeSize();
+            }
+            // also subtract the configured free disk space threshold
+            leastAvailableDiskSpaceBytes -= getFreeBytesThreshold(leastAvailablePath.getTotal(), highStageWatermark, highStageMaxHeadroom)
+                .getBytes();
+            // the rest is the maximum disk space available for a new merge task
+            long maxMergeSizeLimit = Math.max(0L, leastAvailableDiskSpaceBytes);
+            queuedMergeTasks.updateMaxPriorityLimit(maxMergeSizeLimit);
+        }
+
+        private static ByteSizeValue getFreeBytesThreshold(
+            ByteSizeValue total,
+            RelativeByteSizeValue watermark,
+            ByteSizeValue maxHeadroom
+        ) {
+            // If bytes are given, they can be readily returned as free bytes. If percentages are given, we need to calculate the free
+            // bytes.
+            if (watermark.isAbsolute()) {
+                return watermark.getAbsolute();
+            }
+            return ByteSizeValue.subtract(total, watermark.calculateValue(total, maxHeadroom));
+        }
+    }
+
+    static class PriorityBlockingQueueWithMaxLimit<E> {
+        private final ToLongFunction<? super E> priorityFunction;
+        private final PriorityQueue<E> priorityQueue;
+        private final ReentrantLock lock;
+        private final Condition elementAvailable;
+        private long maxPriorityLimit;
+
+        PriorityBlockingQueueWithMaxLimit(ToLongFunction<? super E> priorityFunction, long maxPriorityLimit) {
+            this.priorityFunction = priorityFunction;
+            this.priorityQueue = new PriorityQueue<E>(64, Comparator.comparingLong(priorityFunction));
+            this.lock = new ReentrantLock();
+            this.elementAvailable = lock.newCondition();
+            this.maxPriorityLimit = maxPriorityLimit;
+        }
+
+        boolean add(E e) {
+            final ReentrantLock lock = this.lock;
+            lock.lock();
+            try {
+                priorityQueue.offer(e);
+                elementAvailable.signal();
+            } finally {
+                lock.unlock();
+            }
+            return true;
+        }
+
+        E take() throws InterruptedException {
+            final ReentrantLock lock = this.lock;
+            lock.lockInterruptibly();
+            E peek;
+            try {
+                while ((peek = priorityQueue.peek()) == null || priorityFunction.applyAsLong(peek) > maxPriorityLimit)
```

Comment:
The `take` method is special here.
In addition to blocking if there's no element, it will also block if the smallest element in the heap is larger than a limit.